File: samples/templatematching_cuda/java/TemplateMatchingCuda.java
import org.bytedeco.javacpp.Loader;
import org.bytedeco.opencv.opencv_java;
import org.bytedeco.opencv.global.opencv_core;
import org.bytedeco.opencv.global.opencv_cudaimgproc;
import org.bytedeco.opencv.global.opencv_imgcodecs;
import org.bytedeco.opencv.global.opencv_imgproc;
import org.bytedeco.opencv.opencv_core.GpuMat;
import org.bytedeco.opencv.opencv_core.Mat;
import org.bytedeco.opencv.opencv_core.Point;
import org.bytedeco.opencv.opencv_core.Scalar;
import org.bytedeco.opencv.opencv_cudaimgproc.TemplateMatching;

public class TemplateMatchingCuda {
    public static void main(String[] args) {
        Loader.load(opencv_java.class);
        Mat img = opencv_imgcodecs.imread("messi5.jpg");
        Mat template = opencv_imgcodecs.imread("template.jpg");
        int width = template.cols();
        int height = template.rows();

        TemplateMatching templateMatchingCuda = opencv_cudaimgproc.createTemplateMatching(template.type(), opencv_imgproc.TM_SQDIFF);
        GpuMat img_cuda = new GpuMat(img);
        GpuMat template_cuda = new GpuMat(template);
        GpuMat res_cuda = new GpuMat();
        templateMatchingCuda.match(img_cuda, template_cuda, res_cuda);
        Mat res = new Mat();
        res_cuda.download(res);

        Point top_left = new Point();
        opencv_core.minMaxLoc(res, (double[]) null, null, top_left, null, new Mat());
        Point bottom_right = new Point(top_left.x() + width, top_left.y() + height);
        opencv_imgproc.rectangle(img, top_left, bottom_right, new Scalar(255, 139, 0, 0), 2, opencv_imgproc.CV_AA, 0);
        OpenCVFX.imshow("TemplateMatchingCuda", img);
        while (OpenCVFX.waitKey() != 27) {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        OpenCVFX.destroyAllWindows();
    }

}

File: samples/templatematching_cuda/cpp/src/templatematching_cuda.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <opencv2/cudaimgproc.hpp>

int main(int argc, char *argv[])
{
    cv::Mat img = cv::imread("messi5.jpg");
    cv::Mat template_img = cv::imread("template.jpg");
    int width = template_img.cols;
    int height = template_img.rows;

    cv::Ptr<cv::cuda::TemplateMatching> templateMatchingCuda = cv::cuda::createTemplateMatching(template_img.type(), cv::TM_SQDIFF);

    cv::cuda::GpuMat img_cuda(img);
    cv::cuda::GpuMat template_cuda(template_img);
    cv::cuda::GpuMat res_cuda;
    templateMatchingCuda->match(img_cuda, template_cuda, res_cuda);
    cv::Mat res;
    res_cuda.download(res);

    cv::Point top_left;
    cv::minMaxLoc(res, NULL, NULL, &top_left, NULL);
    cv::Point bottom_right(top_left.x + width, top_left.y + height);

    cv::rectangle(img, top_left, bottom_right, cv::Scalar(255, 139, 0), 2);
    cv::imshow("TemplateMatching", img);
    while (cv::waitKey(1000) != 27)
    {
        continue;
    }
    return 0;
}
